src/js/background/sync.js

```diff
@@ -0,0 +1,461 @@
+const SYNC_DEBUG = true;
+
+const sync = {
+  storageArea: {
+    area: browser.storage.sync,
+
+    async get(){
+      return await this.area.get();
+    },
+
+    async set(options) {
+      return await this.area.set(options);
+    },
+
+    async getDeletedIdentityList() {
+      const storedArray = await this.getStoredItem("deletedIdentityList");
+      return (storedArray) ?  storedArray : [];
+    },
+
+    async getIdentities() {
+      const storedArray = await this.getStoredItem("identities");
+      return (storedArray) ?  storedArray : [];
+    },
+
+    async getDeletedSiteList() { 
+      const storedArray = await this.getStoredItem("deletedSiteList");
+      return (storedArray) ?  storedArray : [];
+    },
+
+    async getCookieStoreIDMap() {
+      const storedArray = await this.getStoredItem("cookieStoreIDmap");
+      return (storedArray) ?  storedArray : {};
+    },
+
+    async getAssignedSites() {
+      const storedArray = await this.getStoredItem("assignedSites");
+      return (storedArray) ?  storedArray : {};
+    },
+
+    async getStoredItem(objectKey) {
+      const outputObject = await this.get(objectKey);
+      if (outputObject && outputObject[objectKey]) 
+        return outputObject[objectKey];
+      return false;
+    },
+
+    async hasSyncStorage(){
+      const inSync = await this.get();
+      return !(Object.entries(inSync).length === 0);
+    },
+
+    async backup(options) {
+      // remove listeners to avoid an infinite loop!
+      await sync.checkForListenersMaybeRemove();
+
+      await updateSyncIdentities();
+      await updateCookieStoreIdMap();
+      await updateSyncSiteAssignments();
+      if (options && options.uuid) 
+        await updateDeletedIdentityList(options.uuid);
+      if (options && options.siteStoreKey) 
+        await addToDeletedSitesList(options.siteStoreKey);
+      if (options && options.undelete) 
+        await removeFromDeletedSitesList(options.undelete);
+      // if (SYNC_DEBUG) {
+      //   const storage = await sync.storageArea.get();
+      //   console.log("inSync: ", storage);
+      //   const localStorage = await browser.storage.local.get();
+      //   console.log("inLocal:", localStorage);
+      //   console.log("idents: ", await browser.contextualIdentities.query({}));
+      // }
+      console.log("Backed up!");
+      await sync.checkForListenersMaybeAdd();
```

Comment:
Do all of these await's need to be done in order? If not push them onto an array and `await Promise.all(arrayname);` whenever you need to ensure it's happened.